File: 392.判断子序列.cpp
// @before-stub-for-debug-begin
// #include <vector>
// #include <string>
// #include "commoncppproblem392.h"

// using namespace std;
// @before-stub-for-debug-end

/*
 * @lc app=leetcode.cn id=392 lang=cpp
 *
 * [392] 判断子序列
 */
#include<vector>
#include<iostream>
#include<limits.h>
#include<string>
#include<algorithm>
using namespace std;
// @lc code=start
class Solution {
public:
    int left_bound(vector<int>& nums, int target) {
        int lo = 0, hi = nums.size() - 1;
        while (lo <= hi) {
            int mid = (lo + hi) >> 1;
            if (nums[mid] < target)
                lo = mid + 1;
            else if (nums[mid] > target)
                hi = mid - 1;
            else
                hi = mid - 1;
        }
        if (lo == nums.size())
            return -1;
        return nums[lo];
    }

    bool isSubsequence(string s, string t) {
        // A65 a97
        int m = s.size(), n = t.size();
        vector<vector<int>> index(59, vector<int>());
        for (int i = 0; i < n; i++) {
            index[t[i] - 'A'].emplace_back(i);
        }
        int j = 0;
        for (int i = 0; i < m; i++) {
            char c = s[i];
            if (index[c - 'A'].empty())
                return false;
            int search = left_bound(index[c - 'A'], j);
            if (search == -1)
                return false;
            j = search + 1;
        }
        return true;
    }
};
// @lc code=end

File: 392.判断子序列.java

/*
 * @lc app=leetcode.cn id=392 lang=java
 *
 * [392] 判断子序列
 */
import java.util.*;

// @lc code=start
class Solution {
    int left_bound(ArrayList<Integer> arr, int target) {
        int left = 0, right = arr.size();
        while (left < right) {
            int mid = left + (right - left) / 2;
            if (target > arr.get(mid)) {
                left = mid + 1;
            } else {
                right = mid;
            }
        }
        if (left == arr.size()) {
            return -1;
        }
        return left;
    }

    public boolean isSubsequence(String s, String t) {
        ArrayList<Integer>[] index = new ArrayList[26];
        int m = s.length(), n = t.length();
        int i = 0, j = 0;
        for (int k = 0; k < n; k++) {
            if (index[t.charAt(k) - 'a'] == null) {
                index[t.charAt(k) - 'a'] = new ArrayList<>();
            }
            index[t.charAt(k) - 'a'].add(k);
        }
        while (i < m && j < n) {
            char c = s.charAt(i);
            ArrayList<Integer> indexes = index[c - 'a'];
            if (indexes == null)
                return false;
            int target = left_bound(indexes, j);
            if (target != -1) {
                j = indexes.get(target) + 1;
                i++;
            } else {
                return false;
            }
        }
        return i == m;
    }
}
// @lc code=end
